### PR TITLE
Dragging OutputAreaWidget produces new OutputAreaWidget mirroring same model

### PR DESCRIPTION
**Pull request by danielballan**
Drag-and-drop any notebook output areas to generate a new stand-alone widget that mirrors that output area.

Thanks for the hand, @sccolbert.

output_dragdrop

### CODE REVIEW

**Comment by danielballan** [comment]
If the origin cell is deleted, one is left with an empty mirror widget. What is the proper way to detect that the model is dead and dispose of the extra widget?

**Comment by jasongrout** [comment]
@blink1073 and I were talking about this. A model could, for example, have a signal that it is being disposed. A widget could listen to that signal and dispose itself when the model is being disposed. Then it would be on the owner of a widget, if it is changing the widget's model, to change the model before disposing the old model. @sccolbert, what do you think?

**Comment by danielballan** [comment]
As written, this breaks slider widgets. The slider sees the mousemove event, but it propagates it to the OutputWidgetArea, and a Drag is started. Should jupyter-js-widgets block propagation of events to parents?

**Comment by sccolbert** [comment]
We should probably be more selective on *which* parts of the output area are valid for starting a drag. Since an output area can contain arbitrary output, we should probably *not* consider rendered output as a viable drag-start target. Maybe restrict the target area to the left side gutter?

**Comment by danielballan** [comment]
@sccolbert That makes sense to me. See new commit, defining a widget class, `OutputGutter`, and moving the `_startDrag` logic there.

**Comment by sccolbert** [comment]
@danielballan can you post a gif of your most recent commit? If you need a tool to record it, most all of us use this one: http://www.cockos.com/licecap/

**Comment by SylvainCorlay** [comment]
This is awesome. Adding a semantic class name to the mirrored output would help styling it differently (for example not display the top-level prompt in the mirrored output).

**Comment by danielballan** [comment]
Sure. This recording demonstrates two things that work and two things that don't.

Good:
1. Dragging from the left-side "gutter" of an output area generates a "mirrored" copy that watches the same model.
2. Dragging from anywhere else in the main output area now does nothing, which avoids interfering with IPython widgets or other interactive elements there.

Bad:
3. Deleting a cell leaves any extant mirror widget(s) blank.
4. IPython widgets do not appear in mirror widgets.

demo-424-1

Addressing (3), I wrote (unpushed) code in line with @jasongrout's suggestion above. When an OutputAreaModel is disposed, a new Signal will cause the corresponding OutputAreaWidget to be disposed. But I'm not sure what signal to watch to get this process started. Contrary to my expectation, deleting a cell does not seem to dispose that cell's OutputAreaModel.

I don't know where to start with (4). Is the problem obvious to anyone?

**Comment by danielballan** [comment]
I think I solved (3) cleanly. See latest commit and this screencap of it working. I still could use help with (4).

demo-424-2

**Comment by blink1073** [comment]
@danielballan, you need to explicitly set the "trusted" attribute of the mirrored widget, the "widget" mimedata is being stripped.

**Comment by danielballan** [comment]
Thanks, @blink1073. That did it.

So, this is now working and ready for review / discussion about whether it is a good idea. :- )

demo-424-3

**Comment by sccolbert** [comment]
The other point of discussion (and maybe we want to handle this in a separate PR) is the UI/UX to indicate the output drag target area. As it currently sits, the draggability (is that a word) of the output area is somewhat undiscoverable. We may want to indicate it visually somehow, along with perhaps making a command to "mirror" the output of the selected cell in a new pane (maybe also another PR).

cc @ellisonbg for UI/UX input

**Comment by sccolbert** [comment]
Also, since it may not be apparent from my typing: THIS.IS.AWESOME!

**Comment by danielballan** [comment]
Haha, thanks @sccolbert. All of the above sound reasonable to me -- including "draggability" as a word.

File: src/notebook/output-area/model.ts [hunk 1]
```diff
@@ -43,6 +43,14 @@ class OutputAreaModel implements IDisposable {
   }
 
   /**
+   * A signal emitted when the model is disposed.
+   */
+  get disposed(): ISignal<OutputAreaModel, void> {
+    console.log('get model disposed signal')
```
Comment (blink1073) [inline src/notebook/output-area/model.ts]: Please remove debug print.

File: src/notebook/output-area/widget.ts [hunk 2]
```diff
@@ -434,6 +484,149 @@ namespace OutputAreaWidget {
...
+      break;
+    }
+  }
+
+  /**
+   * A message handler invoked on an `'after-attach'` message.
+   */
+  protected onAfterAttach(msg: Message): void {
+    super.onAfterAttach(msg);
+    let node = this.node;
```
Comment (blink1073) [inline src/notebook/output-area/widget.ts]: You don't need the temp variable here, you can write `this.node.addEventListener`).

File: src/notebook/output-area/widget.ts [hunk 3]
```diff
@@ -434,6 +484,149 @@ namespace OutputAreaWidget {
...
+    super.onBeforeDetach(msg);
+    let node = this.node;
+    node.removeEventListener('mousedown', this);
+  }
+
+  /**
+   * Handle the `'mousedown'` event for the widget.
+   */
+  private _evtMousedown(event: MouseEvent): void {
+
```
Comment (blink1073) [inline src/notebook/output-area/widget.ts]: Extraneous line.

File: src/notebook/output-area/widget.ts [hunk 4]
```diff
@@ -434,6 +484,149 @@ namespace OutputAreaWidget {
...
+    }
+    this._dragData = null;
+    this._drag = null;
+    this._selectTimer = null;
+    super.dispose();
+  }
+
+  private _drag: Drag = null;
+  private _dragData: { pressX: number, pressY: number } = null;
+  private _selectTimer = -1;
```
Comment (blink1073) [inline src/notebook/output-area/widget.ts]: It looks like this is unused, was it meant for something?

**Comment by blink1073** [comment]
Great first PR, @danielballan, you nailed all of our design patterns.

File: src/notebook/output-area/widget.ts (hunk 4 above)
Comment (danielballan) [inline src/notebook/output-area/widget.ts]: Ah, this is no longer relevant. I'll remove this stuff.

**Comment by blink1073** [comment]
Awesome, I added a demo GIF to your top level comment, will merge once Travis agrees.

**Pull request merged**